common/src/graph_collision.rs

```diff
@@ -12,39 +16,121 @@ use crate::{
-            voxels: ref voxel_data,
-            ..
-        } = graph[chunk] else {
-            // Collision checking on unpopulated chunk
-            panic!("Collision checking on unpopulated chunk");
-        };
-    let local_ray = chunk.vertex.node_to_dual().cast::<f32>() * node_transform * ray;
-
-    // Check collision within a single chunk
-    chunk_sphere_cast(
-        collider_radius,
-        voxel_data,
-        layout,
-        &local_ray,
-        tanh_distance,
-    )
-    .map(|hit| GraphCastHit {
-        tanh_distance: hit.tanh_distance,
-        chunk,
-        normal: math::mtranspose(&node_transform) * chunk.vertex.dual_to_node().cast() * hit.normal,
-    })
+) -> Result<Option<GraphCastHit>, SphereCastError> {
+    // A collision check is assumed to be a miss until a collision is found.
+    // This `hit` variable gets updated over time before being returned.
+    let mut hit: Option<GraphCastHit> = None;
+
+    // Start a breadth-first search of the graph's chunks, performing collision checks in each relevant chunk.
+    // The `chunk_queue` contains ordered pairs containing the `ChunkId` and the transformation needed to switch
+    // from the original node coordinates to the current chunk's node coordinates.
+    let mut visited_chunks = FxHashSet::<ChunkId>::default();
+    let mut chunk_queue: VecDeque<(ChunkId, na::Matrix4<f32>)> = VecDeque::new();
+    chunk_queue.push_back((ChunkId::new(position.node, Vertex::A), position.local));
+
+    // Precalculate the chunk boundaries for collision purposes. If the collider goes outside these bounds,
+    // the corresponding neighboring chunk will also be used for collision checking.
+    let klein_lower_boundary = collider_radius.tanh();
+    let klein_upper_boundary =
+        ((Vertex::chunk_to_dual_factor() as f32).atanh() - collider_radius).tanh();
+
+    // Breadth-first search loop
+    while let Some((chunk, node_transform)) = chunk_queue.pop_front() {
+        let Chunk::Populated {
+                voxels: ref voxel_data,
+                ..
+            } = graph[chunk] else {
+                // Collision checking on unpopulated chunk
+                return Err(SphereCastError::OutOfBounds);
+            };
+        let local_ray = chunk.vertex.node_to_dual().cast::<f32>() * node_transform * ray;
+
+        // Check collision within a single chunk
+        let current_tanh_distance = hit.as_ref().map_or(tanh_distance, |hit| hit.tanh_distance);
+        hit = chunk_sphere_cast(
+            collider_radius,
+            voxel_data,
+            layout,
+            &local_ray,
+            current_tanh_distance,
+        )
+        .map_or(hit, |hit| {
+            Some(GraphCastHit {
+                tanh_distance: hit.tanh_distance,
+                chunk,
+                normal: math::mtranspose(&node_transform)
+                    * chunk.vertex.dual_to_node().cast()
+                    * hit.normal,
+            })
+        });
```

Comment:
Shouldn't this only replace `hit` if the newly detected hit is closer? Consider e.g. a spherecast roughly parallel to and overlapping wit the edge of a chunk. This would be a good test case, too.

Reply:
That is handled by the fact that `chunk_sphere_cast` only detects collisions at `current_tanh_distance` or closer. I agree with adding a test case for this, though.

Reply:
Oh, that's clever, and obvious in retrospect. Didn't read closely enough I guess!

Reply:
I went ahead and added the test case.